crates/symbolicator-service/src/services/symbolication/sourcemap.rs

```diff
@@ -0,0 +1,451 @@
+{
+    dist: null,
+    name: '~/assets/bundle.min.js.map',
+    id: 42,
+    headers: {}
+}
+
+// Create SourceMapCache from the source and sourcemap files
+let smcache_writer = SourceMapCacheWriter::new(source_file, sourcemap_file).unwrap();
+let mut smcache_buf = vec![];
+smcache_writer.serialize(&mut smcache_buf).unwrap();
+
+// Cache the SourceMapCache using `sha1` of the abs_path artifact as the key `87324`
+
+// Process JS frame.
+
+*/
+
+pub type OwnedSourceMapCache = SelfCell<ByteView<'static>, SourceMapCache<'static>>;
+
+fn smcache_from_files(
+    source: &mut NamedTempFile,
+    sourcemap: &mut NamedTempFile,
+) -> OwnedSourceMapCache {
+    use std::io::Read;
+
+    let mut source_buf = String::new();
+    source.as_file_mut().read_to_string(&mut source_buf).ok();
+
+    let mut sourcemap_buf = String::new();
+    sourcemap
+        .as_file_mut()
+        .read_to_string(&mut sourcemap_buf)
+        .ok();
+
+    let smcache_writer = SourceMapCacheWriter::new(&source_buf, &sourcemap_buf).unwrap();
+    let mut smcache_buf = vec![];
+    smcache_writer.serialize(&mut smcache_buf).unwrap();
+
+    let byteview = ByteView::from_vec(smcache_buf);
+    SelfCell::try_new::<symbolic::sourcemapcache::SourceMapCacheError, _>(byteview, |data| unsafe {
+        SourceMapCache::parse(&*data)
+    })
+    .unwrap()
+}
+
+#[derive(Default, Debug)]
+struct FrameSourceMapping {
+    pub source_file: Option<NamedTempFile>,
+    pub source_artifact: Option<SearchArtifactResult>,
+
+    pub sourcemap_file: Option<NamedTempFile>,
+    pub sourcemap_artifact: Option<SearchArtifactResult>,
+
+    pub smcache: Option<OwnedSourceMapCache>,
+}
+
+impl SymbolicationActor {
+    async fn collect_stacktrace_artifacts(
+        &self,
+        request: &JsProcessingSymbolicateStacktraces,
+    ) -> HashMap<String, FrameSourceMapping> {
+        let mut unique_abs_paths: HashSet<_> = HashSet::new();
+        for stacktrace in &request.stacktraces {
+            for frame in &stacktrace.frames {
+                unique_abs_paths.insert(frame.abs_path.clone());
+            }
+        }
+
+        let mut collected_artifacts = HashMap::new();
+        let release_archive = self.sourcemaps.list_artifacts(request.source.clone()).await;
+
+        for abs_path in unique_abs_paths {
+            let mut frame_source_mapping = FrameSourceMapping {
+                ..Default::default()
+            };
+
+            for candidate in get_release_file_candidate_urls(&abs_path) {
+                if let Some(source_artifact) = release_archive.get(&candidate) {
+                    if let Some(source_file) = self
+                        .sourcemaps
+                        .fetch_artifact(SentryRemoteFile::new(
+                            request.source.clone(),
+                            source_artifact.id.clone(),
+                            SentryFileType::ReleaseArtifact,
+                        ))
+                        .await
```

Comment:
this await is "blocking" the loop here, same for the sourcemap reference below.
Ideally we would `map` the set of `unique_abs_paths` into a list of futures for fetching source files concurrently.

but we can also leave this for later, as this is the core of the caching layer we have to figure out.